File: packages/gizmos-env/lib/Gizmo.ts
import {
  type GizmoLevel,
  type EnergyWithAny,
  type Energy,
  type AllGizmoLevel,
  BuildMethod,
  Stage,
} from './common'
import type { Player } from './Player'

export enum GizmoType {
  PICK = 'PICK',
  BUILD = 'BUILD',
  UPGRADE = 'UPGRADE',
  CONVERTER = 'CONVERTER',
  FILE = 'FILE',
}
export type Effect =
  | { type: 'na' }
  | { type: 'free_draw'; num: number }
  | { type: 'free_pick'; num: number }
  | { type: 'add_point_token'; num: number }
  | { type: 'extra_file' }
  | { type: 'extra_research' }
  | { type: 'extra_build'; level: GizmoLevel[] }
  | { type: 'energy_as_point' }
  | { type: 'token_as_point' }
export interface GizmoBasic<Level = GizmoLevel> {
  id: number
  level: Level
  energy_type: EnergyWithAny
  energy_cost: number
  value: number
  effect?: Effect
}
export interface GizmoInfo {
  id: number
  active: boolean
  used: boolean
  where: 'unknown' | 'board' | 'research' | 'file' | 'player'
  belongs_to: number | null
}
export abstract class Gizmo<Level = GizmoLevel> {
  abstract type: GizmoType
  id: number
  level: Level
  energy_type: EnergyWithAny
  energy_cost: number
  value: number
  effect: Effect

  active: boolean
  used: boolean

  where: 'excluded' | 'pool' | 'board' | 'research' | 'file' | 'player'
  belongs_to: number | null

  assert_available() {
    if (!this.active) {
      throw new Error('gizmo not activated')
    }
    if (this.used) {
      throw new Error('gizmo used')
    }
  }

  used_by(player: Player) {
    this.assert_available()
    switch (this.effect.type) {
      case 'free_draw':
        this.free_draw(player, this.effect.num)
        break
      case 'free_pick':
        this.free_pick(player, this.effect.num)
        break
      case 'add_point_token':
        this.add_point_token(player, this.effect.num)
        break
      case 'extra_file':
        this.extra_file(player)
        break
      case 'extra_research':
        this.extra_research(player)
        break
      case 'extra_build':
        this.extra_build(player, this.effect.level)
        break
      default:
    }
    this.active = false
    this.used = true
  }

  free_draw(player: Player, num: number) {
    if (player.env.energy_pool_len() <= 0) return
    const draw_num = Math.min(
      num,
      player.max_energy_num - player.total_energy_num
    )
    player.env.draw_energy_from_pool(draw_num).forEach(player.add_energy)
  }

  free_pick(player: Player, num: number) {
    if (player.env.state.energy_board.length <= 0) return
    player.env.state.free_pick_num = num
  }

  add_point_token(player: Player, num: number) {
    player.point_token += num
  }

  extra_file(player: Player) {
    player.env.state.curr_stage === Stage.EXTRA_FILE
  }

  extra_research(player: Player) {
    if (player.research_num <= 0) return
    player.env.state.curr_stage = Stage.EXTRA_RESEARCH
  }

  extra_build(player: Player, level: GizmoLevel[]) {
    player.env.state.free_build = { level }
    player.env.state.curr_stage = Stage.EXTRA_BUILD
  }

  reset_used() {
    this.active = false
    this.used = false
  }

  reset() {
    this.reset_used()
    this.where = 'excluded'
    this.belongs_to = null
  }

  get_value(player: Player) {
    switch (this.effect.type) {
      case 'token_as_point':
        return player.point_token
      case 'energy_as_point':
        return player.total_energy_num
      default:
        return this.value
    }
  }

  get info(): GizmoInfo {
    return {
      id: this.id,
      active: this.active,
      used: this.used,
      where:
        this.where === 'excluded' || this.where === 'pool'
          ? 'unknown'
          : this.where,
      belongs_to: this.belongs_to,
    }
  }

  constructor({
    id,
    level,
    energy_type,
    energy_cost,
    value,
    effect,
  }: GizmoBasic<Level>) {
    this.id = id
    this.level = level
    this.energy_type = energy_type
    this.energy_cost = energy_cost
    this.value = value
    this.effect = effect ?? { type: 'na' }

    this.active = false
    this.used = false

    this.where = 'excluded'
    this.belongs_to = null
  }
}

type GizmoPick = {
  when_pick: Energy[]
}
export class PickGizmo extends Gizmo {
  type = GizmoType.PICK
  when_pick: Energy[]

  is_satisfied(energy: Energy) {
    return this.when_pick.includes(energy)
  }

  on_pick(energy: Energy) {
    if (!this.is_satisfied(energy)) {
      return
    }
    if (!this.used) {
      this.active = true
    }
  }

  constructor({ when_pick, ...basic }: GizmoBasic & GizmoPick) {
    super(basic)
    this.when_pick = when_pick
  }
}

type GizmoBuild = {
  when_build: {
    energy: Energy[] | 'any'
    level: AllGizmoLevel[] | 'any'
    method: BuildMethod[] | 'any'
  }
}
export class BuildGizmo extends Gizmo {
  type = GizmoType.BUILD
  when_build: {
    energy: Energy[] | 'any'
    level: AllGizmoLevel[] | 'any'
    method: BuildMethod[] | 'any'
  }

  is_satisfied(
    player: Player,
    level: GizmoLevel,
    energy: EnergyWithAny,
    method: BuildMethod
  ) {
    return (
      (this.when_build.level === 'any' ||
        this.when_build.level.includes(level)) &&
      (this.when_build.energy === 'any' ||
        energy === 'any' ||
        this.when_build.energy.includes(energy)) &&
      (this.when_build.method === 'any' ||
        this.when_build.method.includes(method)) &&
      (this.effect.type !== 'extra_file' || player.can_file) &&
      (this.effect.type !== 'extra_research' || player.can_research)
    )
  }

  on_build(
    player: Player,
    level: GizmoLevel,
    energy: EnergyWithAny,
    method: BuildMethod
  ) {
    if (!this.is_satisfied(player, level, energy, method)) {
      return
    }
    if (!this.used) {
      this.active = true
    }
  }

  constructor({ when_build, ...basic }: GizmoBasic & GizmoBuild) {
    super(basic)
    this.when_build = when_build
  }
}

type GizmoUpgrade = {
  max_energy_num?: number
  max_file_num?: number
  research_num?: number
  build_from_filed_cost_reduction?: number
  build_from_research_cost_reduction?: number
}
export class UpgradeGizmo extends Gizmo {
  type = GizmoType.UPGRADE
  max_energy_num: number
  max_file_num: number
  research_num: number
  build_from_filed_cost_reduction: number
  build_from_research_cost_reduction: number

  constructor({
    max_energy_num,
    max_file_num,
    research_num,
    build_from_filed_cost_reduction,
    build_from_research_cost_reduction,
    ...basic
  }: GizmoBasic & GizmoUpgrade) {
    super(basic)
    this.max_energy_num = max_energy_num ?? 0
    this.max_file_num = max_file_num ?? 0
    this.research_num = research_num ?? 0
    this.build_from_filed_cost_reduction = build_from_filed_cost_reduction ?? 0
    this.build_from_research_cost_reduction =
      build_from_research_cost_reduction ?? 0
  }
}

export type ConverterFormula<EF = EnergyWithAny, ET = EnergyWithAny> = {
  from: {
    energy: EF
    num: number
  }
  to: {
    energy: ET
    num: number
  }
}
type GizmoConverter = {
  prerequisite?: {
    level: AllGizmoLevel[]
  }
  formulae: ConverterFormula[]
}
export class ConverterGizmo extends Gizmo {
  type = GizmoType.CONVERTER
  prerequisite?: { level: AllGizmoLevel[] }
  formulae: ConverterFormula[]

  is_satisfied(gizmo: Gizmo) {
    return (
      !this.used &&
      (!this.prerequisite || this.prerequisite.level.includes(gizmo.level))
    )
  }

  on_convert(gizmo: Gizmo) {
    if (!this.is_satisfied(gizmo)) return
    this.active = true
  }

  constructor({
    prerequisite,
    formulae,
    ...basic
  }: GizmoBasic & GizmoConverter) {
    super(basic)
    this.prerequisite = prerequisite
    this.formulae = formulae
  }
}

// eslint-disable-next-line @typescript-eslint/ban-types
type GizmoFile = {}
export class FileGizmo<Level = GizmoLevel> extends Gizmo<Level> {
  type = GizmoType.FILE

  on_file() {
    if (!this.used) {
      this.active = true
    }
  }
  constructor({ ...basic }: GizmoBasic<Level> & GizmoFile) {
    super(basic)
  }
}

export function is_upgrade_gizmo(
  gizmo: Gizmo<AllGizmoLevel>
): gizmo is UpgradeGizmo {
  return gizmo instanceof UpgradeGizmo
}
export function is_converter_gizmo(
  gizmo: Gizmo<AllGizmoLevel>
): gizmo is ConverterGizmo {
  return gizmo instanceof ConverterGizmo
}
export function is_pick_gizmo(gizmo: Gizmo<AllGizmoLevel>): gizmo is PickGizmo {
  return gizmo instanceof PickGizmo
}
export function is_build_gizmo(
  gizmo: Gizmo<AllGizmoLevel>
): gizmo is BuildGizmo {
  return gizmo instanceof BuildGizmo
}
export function is_file_gizmo(gizmo: Gizmo<AllGizmoLevel>): gizmo is FileGizmo {
  return gizmo instanceof FileGizmo
}

File: py/Gizmo.py
from __future__ import annotations
from enum import Enum
from typing import Literal, TypeGuard, TypedDict, TYPE_CHECKING

from common import GizmoLevel, EnergyWithAny, Energy, AllGizmoLevel, BuildMethod, Stage

if TYPE_CHECKING:
    from Player import Player


class GizmoType(str, Enum):
    PICK = 'PICK'
    BUILD = 'BUILD'
    UPGRADE = 'UPGRADE'
    CONVERTER = 'CONVERTER'
    FILE = 'FILE'


class NaEffect(TypedDict):
    type: Literal['na']


class FreeDrawEffect(TypedDict):
    type: Literal['free_draw']
    num: int


class FreePickEffect(TypedDict):
    type: Literal['free_pick']
    num: int


class AddPointTokenEffect(TypedDict):
    type: Literal['add_point_token']
    num: int


class ExtraFileEffect(TypedDict):
    type: Literal['extra_file']


class ExtraResearchEffect(TypedDict):
    type: Literal['extra_research']


class ExtraBuildEffect(TypedDict):
    type: Literal['extra_build']
    level: list[GizmoLevel]


class EnergyAsPointEffect(TypedDict):
    type: Literal['energy_as_point']


class TokenAsPointEffect(TypedDict):
    type: Literal['token_as_point']


Effect = NaEffect | FreeDrawEffect | FreePickEffect | AddPointTokenEffect | ExtraFileEffect | ExtraResearchEffect | ExtraBuildEffect | EnergyAsPointEffect | TokenAsPointEffect


class GizmoBasic(TypedDict):
    id: int
    level: AllGizmoLevel
    energy_type: EnergyWithAny
    energy_cost: int
    value: int
    effect: Effect | None


class GizmoInfo(TypedDict):
    id: int
    active: bool
    used: bool
    where: Literal['excluded', 'pool', 'board', 'research', 'file', 'player']
    belongs_to: int | None


class Gizmo:
    type: GizmoType
    id: int
    level: AllGizmoLevel
    energy_type: EnergyWithAny
    energy_cost: int
    value: int
    effect: Effect

    active: bool
    used: bool

    where: Literal['excluded', 'pool', 'board', 'research', 'file', 'player']
    belongs_to: int | None

    def assert_available(self):
        if not self.active:
            raise Exception('gizmo not activated')
        if self.used:
            raise Exception('gizmo used')

    def used_by(self, player: Player):
        self.assert_available()
        if self.effect['type'] == 'free_draw':
            self.free_draw(player, self.effect['num'])
        elif self.effect['type'] == 'free_pick':
            self.free_pick(player, self.effect['num'])
        elif self.effect['type'] == 'add_point_token':
            self.add_point_token(player, self.effect['num'])
        elif self.effect['type'] == 'extra_file':
            self.extra_file(player)
        elif self.effect['type'] == 'extra_research':
            self.extra_research(player)
        elif self.effect['type'] == 'extra_build':
            self.extra_build(player, self.effect['level'])
        self.active = False
        self.used = True

    def free_draw(self, player: Player, num: int):
        if player.env.energy_pool_len() <= 0:
            return
        draw_num = min(num, player.max_energy_num - player.total_energy_num)
        for energy in player.env.draw_energy_from_pool(draw_num):
            player.add_energy(energy)

    def free_pick(self, player: Player, num: int):
        if len(player.env.state['energy_board']) <= 0:
            return
        player.env.state['free_pick_num'] = num

    def add_point_token(self, player: Player, num: int):
        player.point_token += num

    def extra_file(self, player: Player):
        player.env.state['curr_stage'] = Stage.EXTRA_FILE

    def extra_research(self, player: Player):
        if player.research_num <= 0:
            return
        player.env.state['curr_stage'] = Stage.EXTRA_RESEARCH

    def extra_build(self, player: Player, level: list[GizmoLevel]):
        player.env.state['free_build'] = {'level': level}
        player.env.state['curr_stage'] = Stage.EXTRA_BUILD

    def reset_used(self):
        self.active = False
        self.used = False

    def reset(self):
        self.reset_used()
        self.where = 'excluded'
        self.belongs_to = None

    def get_value(self, player: Player):
        if self.effect['type'] == 'token_as_point':
            return player.point_token
        if self.effect['type'] == 'energy_as_point':
            return player.total_energy_num
        return self.value

    @property
    def info(self) -> GizmoInfo:
        return {
            'id': self.id,
            'active': self.active,
            'used': self.used,
            'where': self.where,
            'belongs_to': self.belongs_to,
        }

    def __init__(self, **basic: GizmoBasic):
        self.id = basic['id']
        self.level = basic['level']
        self.energy_type = basic['energy_type']
        self.energy_cost = basic['energy_cost']
        self.value = basic['value']
        self.effect = basic.get('effect', {'type': 'na'})

        self.active = False
        self.used = False

        self.where = 'excluded'
        self.belongs_to = None


class GizmoPick(TypedDict):
    when_pick: list[Energy]


class PickGizmo(Gizmo):
    type = GizmoType.PICK
    when_pick: list[Energy]

    def is_satisfied(self, energy: Energy) -> bool:
        return energy in self.when_pick

    def on_pick(self, energy: Energy):
        if not self.is_satisfied(energy):
            return
        if not self.used:
            self.active = True

    def __init__(self, when_pick: list[Energy], **basic: GizmoBasic):
        super().__init__(**basic)
        self.when_pick = when_pick


class WhenBuild(TypedDict):
    energy: list[Energy] | Literal['any']
    level: list[AllGizmoLevel] | Literal['any']
    method: list[BuildMethod] | Literal['any']


class GizmoBuild(TypedDict):
    when_build: WhenBuild


class BuildGizmo(Gizmo):
    type = GizmoType.BUILD
    when_build: WhenBuild

    def is_satisfied(self, player: Player, level: GizmoLevel, energy: EnergyWithAny, method: BuildMethod) -> bool:
        return (
            (self.when_build['level'] == 'any' or level in self.when_build['level']) and
            (self.when_build['energy'] == 'any' or energy == 'any' or energy in self.when_build['energy']) and
            (self.when_build['method'] == 'any' or method in self.when_build['method']) and
            (self.effect['type'] != 'extra_file' or player.can_file) and
            (self.effect['type'] != 'extra_research' or player.can_research)
        )

    def on_build(self, player: Player, level: GizmoLevel, energy: EnergyWithAny, method: BuildMethod):
        if not self.is_satisfied(player, level, energy, method):
            return
        if not self.used:
            self.active = True

    def __init__(self, when_build: WhenBuild, **basic: GizmoBasic):
        super().__init__(**basic)
        self.when_build = when_build


class GizmoUpgrade(TypedDict):
    max_energy_num: int | None
    max_file_num: int | None
    research_num: int | None
    build_from_filed_cost_reduction: int | None
    build_from_research_cost_reduction: int | None


class UpgradeGizmo(Gizmo):
    type = GizmoType.UPGRADE
    max_energy_num: int
    max_file_num: int
    research_num: int
    build_from_filed_cost_reduction: int
    build_from_research_cost_reduction: int

    def __init__(self, **basic: GizmoBasic & GizmoUpgrade):
        super().__init__(**basic)
        self.max_energy_num = basic.get('max_energy_num', 0)
        self.max_file_num = basic.get('max_file_num', 0)
        self.research_num = basic.get('research_num', 0)
        self.build_from_filed_cost_reduction = basic.get(
            'build_from_filed_cost_reduction', 0)
        self.build_from_research_cost_reduction = basic.get(
            'build_from_research_cost_reduction', 0)


class FormulaSide(TypedDict):
    energy: EnergyWithAny
    num: int


ConverterFormula = dict[Literal['from', 'to'], FormulaSide]


class Prerequisite(TypedDict):
    level: list[AllGizmoLevel]


class GizmoConverter(TypedDict):
    prerequisite: Prerequisite | None
    formulae: list[ConverterFormula]


class ConverterGizmo(Gizmo):
    type = GizmoType.CONVERTER
    prerequisite: Prerequisite | None
    formulae: list[ConverterFormula]

    def is_satisfied(self, gizmo: Gizmo) -> bool:
        return (
            not self.used and
            (not self.prerequisite or gizmo.level in self.prerequisite['level'])
        )

    def on_convert(self, gizmo: Gizmo):
        if not self.is_satisfied(gizmo):
            return
        self.active = True

    def __init__(self, **basic: GizmoBasic & GizmoConverter):
        super().__init__(**basic)
        self.prerequisite = basic.get('prerequisite')
        self.formulae = basic['formulae']


class FileGizmo(Gizmo):
    type = GizmoType.FILE

    def on_file(self):
        if not self.used:
            self.active = True

    def __init__(self, **basic: GizmoBasic):
        super().__init__(**basic)


def is_upgrade_gizmo(gizmo: Gizmo) -> TypeGuard[UpgradeGizmo]:
    return isinstance(gizmo, UpgradeGizmo)


def is_converter_gizmo(gizmo: Gizmo) -> TypeGuard[ConverterGizmo]:
    return isinstance(gizmo, ConverterGizmo)


def is_pick_gizmo(gizmo: Gizmo) -> TypeGuard[PickGizmo]:
    return isinstance(gizmo, PickGizmo)


def is_build_gizmo(gizmo: Gizmo) -> TypeGuard[BuildGizmo]:
    return isinstance(gizmo, BuildGizmo)


def is_file_gizmo(gizmo: Gizmo) -> TypeGuard[FileGizmo]:
    return isinstance(gizmo, FileGizmo)
